agent/src/util/discovery_operator.rs

```diff
@@ -0,0 +1,1394 @@
+    /// In a separate task, it calls connects to each discovery handler in the RegisteredDiscoveryHandlerMap
+    /// with the same protocol name as the Configuration (Configuration.protocol.name). Then, it listens for
+    /// updates from the discovery handler on what devices are currently visible to the node.
+    /// Passes this list to a function that updates the InstanceConnectivityStatus of the Configuration's Instances
+    /// or deletes Instance CRs if needed. If a new instance becomes visible that isn't in the Configuration's InstanceMap,
+    /// a DevicePluginService and Instance CRD are created for it, and it is added to the InstanceMap.
+    ///
+    /// It also spawns a task to check whether Offline Instances have exceeded their grace period, in which case it
+    /// deletes the Instance.
+    pub async fn start_discovery(
+        discovery_operator: DiscoveryOperator,
+        new_discovery_handler_sender: broadcast::Sender<String>,
+        stop_all_discovery_sender: broadcast::Sender<()>,
+        finished_all_discovery_sender: &mut broadcast::Sender<()>,
+    ) -> Result<(), Box<dyn std::error::Error + Send + Sync + 'static>> {
+        let config = discovery_operator.get_config();
+        info!(
+            "start_discovery - entered for protocol {}",
+            config.spec.protocol.name
+        );
+        let config_name = config.metadata.name.clone();
+        let mut tasks = Vec::new();
+        let discovery_operator = Arc::new(discovery_operator);
+        let task1_discovery_operator = discovery_operator.clone();
+        // Call discover on already registered Discovery Handlers for this Configuration's protocol
+        tasks.push(tokio::spawn(async move {
+            do_discover(
+                task1_discovery_operator,
+                Arc::new(Box::new(k8s::create_kube_interface())),
+            )
+            .await
+            .unwrap();
+        }));
+        let mut stop_all_discovery_receiver = stop_all_discovery_sender.subscribe();
+        let mut new_discovery_handler_receiver = new_discovery_handler_sender.subscribe();
+        let task2_discovery_operator = discovery_operator.clone();
+        tasks.push(tokio::spawn(async move {
+            let mut inner_tasks = Vec::new();
+            loop {
+                tokio::select! {
+                    _ = try_receive(&mut stop_all_discovery_receiver) => {
+                        trace!("start_discovery - received message to stop discovery for configuration {}", task2_discovery_operator.get_config().metadata.name);
+                        // stop_offline_checks_sender.send(()).unwrap();
+                        task2_discovery_operator.stop_all_discovery().await;
+                        break;
+                    },
+                    result = try_receive(&mut new_discovery_handler_receiver) => {
+                        // check if it is this protocol
```

Comment:
Suggestion: Maybe this can move to a separate fn for better readability

Reply:
Great suggestion! Fixed in 80e02b7.